Merge pre-existing package.json with file generated by Cloverfield

halhenke opened the pull request:
Re issue #23 
- Tried to make it fit the pre-existing code structure as much as possible.
- There is some "intelligence" to the merging from the `package-merge` package rather than a straight object merge - keywords field will be merged without duplicates etc. Might want to take a quick look at that package to see if its desired behaviour: 
  - https://github.com/izaakschroeder/package-merge/blob/master/lib/merge.js

halhenke committed:
require package-merge

halhenke committed:
functions to merge a pre-existing package.son file with generated template and tests

halhenke committed:
Add merge package step between template compilation and content saving

ericelliott commented:
Tests fail due to [a missing package](https://circleci.com/gh/cloverfield-tools/cf-package/47).

halhenke commented:
Yeah, seems the guy who maintains package-merge didnt declare `lodash` as a dependency. 

Its passing on my machine seemingly because its a dependency of `precommit-hook` thats been spat into node_modules by npm 3 as far as i can tell. 

Not sure what the best approach is - theres not a ton of functionality in that package but i thought it would be nice to have. Could fork it and make a PR and in the meantime use the fork or do a simpler 
`{ ...origPkg, ...genPkg}` type merge?

nkbt commented on package.json:
```diff
@@ -54,6 +54,8 @@
     "isparta": "^3.0.4",
     "nsp": "^1.1.0",
+    "package-merge": "^0.1.2",
+    "precommit-hook": "^3.0.0",
```
oops, that was deleted previously

nkbt commented on lib/mergePack.js:
```diff
@@ -0,0 +1,25 @@
+import merge from 'package-merge';
```
You can simply use https://lodash.com/docs#merge which is recursive.

https://www.npmjs.com/package/lodash.merge would work fine.

halhenke replied:
Yeah, happy to do this.

nkbt commented on lib/mergePack.js:
```diff
@@ -0,0 +1,25 @@
+import fs from 'fs';
+import path from 'path';
+
+const getPackage = () => {
```
`export const getPackage`

nkbt commented on lib/mergePack.js:
```diff
@@ -0,0 +1,25 @@
+  }
+  return compiledFiles;
+};
+export {getPackage, mergePackages};
```
better export each function.

halhenke committed:
export functions rather than object in mergePack

halhenke committed:
Swap out package-merge for loads.merge for package.json merging

halhenke committed:
test that package.json output has decent formatting

nkbt commented:
Hi, @halhenke can you please update PR so I can merge it? thanks :ok_hand:

halhenke commented:
OK for sure - sorry thought i already had - must be losing it... :-( 

![](http://gif.co/tOOx.gif)

nkbt commented on package.json:
```diff
@@ -57,6 +58,7 @@
     "husky": "^0.10.1",
     "isparta": "^3.0.4",
     "nsp": "^2.0.2",
+    "package-merge": "^0.1.2",
```
I guess  this one is extra now.

halhenke replied:
Ah - i screwed up the rebase! will fix

nkbt commented on lib/mergePack.js:
```diff
@@ -0,0 +1,24 @@
+    const newPackagePath = path.join(process.cwd(), 'package.json');
+    const packageIndex = destinations.indexOf(newPackagePath);
+    const packageContents = JSON.parse(compiledFiles[packageIndex]);
+    const mergedPack = JSON.stringify(merge(origPackage, packageContents), null, 2);
```
I am not quite sure but I reckon it should be the other way around. package from template should not override fields in already existing package. Only add missing ones.

I mean if I used this template and later changed how my lib is tested, I'd prefer to keep it as it is. But if there are new scripts added, it would be nice to add them to my lib too.

@ericelliott  what is your idea on merging package.json?

halhenke replied:
Yeah I think you are right - and looking back at the original issue that is the described behaviour.

halhenke committed:
Get rid of unused `package.merge` package

halhenke committed:
Swap order of arguments in loads.merge in order to choose users original package fields over template ones

nkbt commented:
Sorry, for the delay, @halhenke 
I've just checked the new generation process. Package.json is merged respecting existing values. Other files not, though. Like README.md has new values, though package.json still has old ones. 

I am not sure about intended behaviour here. I can suggest that when we merge package.json we can see which values have been changed and replace them accordingly in other files too. But not touching values that should not be changed.

Alternative (and much easier) solution could be - read package.json to get "suggested" values for the second run. Then either I just hit "Enter" and use those default ones - or enter new ones - does not matter. Replace all things as usual.

Honestly after playing with it a bit I'd rather go for the second solution.

Related to #44 by the way.

ericelliott commented:
:+1:

halhenke commented:
Hey sorry for delay in getting back to you Nik - busy few days,

Didn't realise we wanted to merge README also - was only thinking of package.json. 

I also like the second idea - at least it makes people aware of whats happening. Will rework it and see if i can come up with something that deals with other files - think there could be some hard to deal with edge cases - but otherwise will go for the user prompt option.

Did you want me to try and work in suggestions from #44 also (i got the impression the other guy didnt want to)? Dont mind.

nkbt commented:
Awesome! #44 looks like duplicate to me then.

halhenke commented:
Cool! :+1:

ericelliott commented:
What's going on with this?

halhenke commented:
Sorry, sort of got lost in Christmas stuff - have a couple of free days, will try to finish it off.

ericelliott commented:
Cool. Thanks!

halhenke committed:
Add object rest/spread syntax

halhenke committed:
Very Rough version of changes

halhenke committed:
use `parse-author` & `parse-github-url` to get author & Github info from pre-existing package.json if available

halhenke committed:
Option to merge pre-existing dependencies, devDependencies and scripts into template package.json

halhenke committed:
Refactor

- change big chain of if else statements into something a little more extensible
- move author/github stuff into separate file

halhenke commented:
OK this thing has been dragging on for a while (apologies) would love to wrap it up

Not entirely happy with what i've done (it feels a bit hack-ish) but wanted to check its the direction we wanted.

Anyway if a `package.json` file is found that exists:
- grab author/email/github from package.json & use as defaults
- read `dependecies`, `devDependencies` and `scripts` fields and add questions whether the user wants to merge them into the template `package.json` (could easily add other fields like peerDependencies if you like)

Feels a bit inelegant because its not using the same handlebars template/vars method to modify most fields - though we do now have author/email variables etc that will now be used in other templates also.

Could use more tests. Anyway just wanted to get any feedback/suggested changes when you guys have time.

halhenke commented:
Hey guys - just wanted to say - if this PR just missed the mark or isnt working out I wont be offended if you just want to kill it or anything. Its all pretty fuzzy for me at this point anyway.

ericelliott commented:
@nkbt Comments?

nkbt commented:
Wow, that looks cool, pretty much package.json-specific merging. I wish github sends emails on PR updated. I'll try it on couple of existing projects and let you know asap!

ericelliott commented:
@nkbt Any more thoughts?

nkbt commented on lib/mergePack.js:
```diff
@@ -0,0 +1,51 @@
+    const newPackagePath = path.join(process.cwd(), 'package.json');
+    const packageIndex = destinations.indexOf(newPackagePath);
+    const packageContents = JSON.parse(compiledFiles[packageIndex]);
+    const mergedPack = JSON.stringify(merge(packageContents, origPackage), null, 2);
```
I've tried it several times and tried to fix it too... No luck. With current implmentation no matter what I chose deps/devdeps/scrips are preserved and never updated. If I swap `packageContents` and `origPackage` - they always updated instead. Somehow those props from `setPackageProps` are just totally ignored =(

halhenke commented:
:disappointed: 

I'll take a look at it either tonight or later this weekend.
